waspls/src/Wasp/Backend/ConcreteParser/Internal.hs

```diff
@@ -0,0 +1,585 @@
+                }
+        pushNode newNode
+        consumeWhitespace
+      | otherwise -> pure ()
+
+-- | Peek the immediate next token in input (including whitespace)
+peek :: ParserM (Maybe Token)
+peek =
+  gets pstateInput >>= \case
+    [] -> return Nothing
+    tok : _ -> return (Just tok)
+
+-- | Advance past the next token in the input, updating offsets.
+advance :: ParserM ()
+advance =
+  gets pstateInput >>= \case
+    [] -> modify (\s -> s {pstateLastOffset = pstateNextOffset s})
+    (consumed : remaining) -> do
+      newLastOffset <- gets pstateNextOffset
+      let newNextOffset = newLastOffset + tokenWidth consumed
+      modify (\s -> s {pstateInput = remaining, pstateLastOffset = newLastOffset, pstateNextOffset = newNextOffset})
+
+-- | @makeError actual expected@ throws an error.
+--
+-- Assumes @actual@ has not been `advance`d past yet (and does not advance past
+-- it).
+makeError :: Maybe Token -> TokenSet -> ParserM ()
+makeError Nothing eset = do
+  offset <- gets pstateNextOffset
+  throwError $ ParseError $ UnexpectedEOF offset eset
+makeError (Just actual) eset = do
+  offset <- gets pstateNextOffset
+  throwError $ ParseError $ Unexpected (Span offset (offset + tokenWidth actual)) (tokenKind actual) eset
+
+-- | Make an error node. Error node should only be produced when catching and
+-- pushing errors to the state.
+errorNode ::
+  -- | error to make an error node from
+  ParseError ->
+  -- | If "True", give the error node a width of 0. Otherwise, use the
+  -- width of the error item.
+  Bool ->
+  SyntaxNode
+errorNode (UnexpectedEOF _ _) _ =
+  SyntaxNode
+    { snodeKind = S.Error,
+      snodeWidth = 0,
+      snodeChildren = []
+    }
+errorNode (Unexpected (Span so eo) _ _) zeroWidth =
+  SyntaxNode
+    { snodeKind = S.Error,
+      snodeWidth = if zeroWidth then 0 else eo - so,
+      snodeChildren = []
+    }
+
+-- | Add a node to the output.
+pushNode :: SyntaxNode -> ParserM ()
+pushNode node = modify (\s -> s {pstateNodes = pstateNodes s ++ [node]})
+
+-- | Add an error to the output (does not throw the error).
+pushError :: ParseError -> ParserM ()
+pushError err = modify (\s -> s {pstateErrors = pstateErrors s ++ [err]})
+
+-- | Add an "ExpectedSet" to the following stack.
+pushFollowing :: TokenSet -> ParserM ()
+pushFollowing eset = modify (\s -> s {pstateFollowing = eset : pstateFollowing s})
+
+-- | Remove the last added "ExpectedSet" from the following stack.
+popFollowing :: ParserM ()
+popFollowing = modify (\s -> s {pstateFollowing = drop 1 $ pstateFollowing s})
+
+-- | Check if a "TokenKind" (or end of file) is one of the first possible tokens
+-- in a "GrammarRule".
+willSucceedIn :: Maybe TokenKind -> GrammarRule -> Bool
+_ `willSucceedIn` Succeed = True
+k `willSucceedIn` grammarRule = k `TokenSet.member` findFirstTokens grammarRule
+
+-- | Find the "ExpectedSet" of "TokenKind"s that the "GrammarRule" would succeed on
+-- as its first input.
+findFirstTokens :: GrammarRule -> TokenSet
```

Comment:
Then maybe call it `getExpectedTokens`? Or `getValidNextTokens`? or `getValidFirstTokens`?